ksqldb-engine/src/main/java/io/confluent/ksql/analyzer/ColumnReferenceExtractor.java

```diff
@@ -0,0 +1,61 @@
+package io.confluent.ksql.analyzer;
+
+import io.confluent.ksql.execution.expression.tree.ComparisonExpression;
+import io.confluent.ksql.execution.expression.tree.Expression;
+import io.confluent.ksql.execution.expression.tree.LogicalBinaryExpression;
+import io.confluent.ksql.execution.expression.tree.QualifiedColumnReferenceExp;
+import io.confluent.ksql.execution.expression.tree.UnqualifiedColumnReferenceExp;
+import io.confluent.ksql.execution.expression.tree.VisitParentExpressionVisitor;
+import io.confluent.ksql.name.ColumnName;
+import java.util.HashSet;
+import java.util.Optional;
+import java.util.Set;
+
+/**
+ * The purpose of this class is to extract ColumnNames from an {@link Expression}.
+ * This currently successfully traverses the expression tree only if
+ * non-leaf nodes are composed exclusively of ComparisonExpressions and LogicalBinaryExpressions,
+ * as this was the behavior necessary to disable pull queries and scalable push queries
+ * with ROWPARTITION and ROWOFFSET pseudocolumns (as these queries do not work with other clauses).
+ */
+public class ColumnReferenceExtractor extends
+    VisitParentExpressionVisitor<Optional<Expression>, Void> {
```

Comment:
Extend `TraversalExpressionVisitor` instead of `VisitParentExpressionVisitor`. Then you won't need to implement methods like `visitComparisonExpression()`, and the functionality of this visitor won't be limited to just comparisons and logical binary expressions. (It's also not true that those are the only two types of expressions we need for pull queries and scalable push queries -- for example, there could be other (arbitrary) types of expressions in the select statements).

Reply:
BTW, while looking at the code just now I saw that https://github.com/confluentinc/ksql/blob/master/ksqldb-execution/src/main/java/io/confluent/ksql/execution/util/ColumnExtractor.java already exists. I think you can just use that directly.